fish-rust/src/wildcard.rs

```diff
@@ -11,3 +48,1249 @@ pub const ANY_STRING_RECURSIVE: char = char_offset(WILDCARD_RESERVED_BASE, 2);
+) -> WString {
+    if let Some(complete_sep_loc) = completion.find_char(PROG_COMPLETE_SEP) {
+        // This completion has an embedded description, do not use the generic description.
+        assert!(
+            completion.len() > complete_sep_loc,
+            "resolve_description lacks a length assumption"
+        );
+        let (comp, description) = completion.split_at(complete_sep_loc + 1);
+        *completion = comp;
+        return description.to_owned();
+    }
+
+    if let Some(f) = description_func {
+        if expand_flags.contains(ExpandFlags::GEN_DESCRIPTIONS) {
+            return f(full_completion);
+        }
+    }
+
+    WString::new()
+}
+
+// A transient parameter pack needed by wildcard_complete.
+struct WcCompletePack<'orig, 'f> {
+    pub orig: &'orig wstr,
+    pub desc_func: Option<&'f dyn Fn(&wstr) -> WString>,
+    pub expand_flags: ExpandFlags,
+}
+
+// Weirdly specific and non-reusable helper function that makes its one call site much clearer.
+fn has_prefix_match(comps: &CompletionReceiver, first: usize) -> bool {
+    comps[first..]
+        .iter()
+        .any(|c| c.r#match.is_exact_or_prefix() && c.r#match.case_fold == CaseFold::samecase)
+}
+
+/// Matches the string against the wildcard, and if the wildcard is a possible completion of the
+/// string, the remainder of the string is inserted into the out vector.
+///
+/// We ignore ANY_STRING_RECURSIVE here. The consequence is that you cannot tab complete **
+/// wildcards. This is historic behavior.
+/// is_first_call is default false.
+#[allow(clippy::unnecessary_unwrap)]
+fn wildcard_complete_internal(
+    s: &wstr,
+    wc: &wstr,
+    params: &WcCompletePack,
+    flags: CompleteFlags,
+    // it is easier to recurse with this over taking it by value
+    mut out: Option<&mut CompletionReceiver>,
+    is_first_call: bool,
+) -> WildcardResult {
+    // Maybe early out for hidden files. We require that the wildcard match these exactly (i.e. a
+    // dot); ANY_STRING not allowed.
+    if is_first_call
+        && !params
+            .expand_flags
+            .contains(ExpandFlags::ALLOW_NONLITERAL_LEADING_DOT)
+        && s.char_at(0) == '.'
+        && wc.char_at(0) != '.'
+    {
+        return WildcardResult::NoMatch;
+    }
+
+    // Locate the next wildcard character position, e.g. ANY_CHAR or ANY_STRING.
+    let next_wc_char_pos = wc
+        .chars()
+        .position(|c| matches!(c, ANY_CHAR | ANY_STRING | ANY_STRING_RECURSIVE));
+
+    // Maybe we have no more wildcards at all. This includes the empty string.
+    if next_wc_char_pos.is_none() {
+        // Try matching
+        let Some(m) = string_fuzzy_match_string(wc, s, false) else {
+            return WildcardResult::NoMatch;
+        };
+
+        // If we're not allowing fuzzy match, then we require a prefix match.
+        let needs_prefix_match = !params.expand_flags.contains(ExpandFlags::FUZZY_MATCH);
+        if needs_prefix_match && m.is_exact_or_prefix() {
```

Comment:
```suggestion
        if needs_prefix_match && !m.is_exact_or_prefix() {
```

I guess I should have done the review before testing (but then I might have missed these)